Raise clear exceptions when scorers can't be registered, support custom scorers with mlflow type hints

dbczumar opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/dbczumar/mlflow/pull/18043?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18043/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18043/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/18043/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Raise clear exceptions when scorers can't be registered, support custom scorers with mlflow type hints

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- Do not modify or remove any text inside the parentheses -->

- [X] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/genai/scorers/base.py:
```diff
@@ -295,17 +295,18 @@ def _reconstruct_decorator_scorer(cls, serialized: SerializedScorer) -> "Scorer"
                 f"{serialized.mlflow_version or 'unknown'}, "
                 f"serialization version: {serialized.serialization_version or 'unknown'}. "
-                f"Current MLflow version: {mlflow.__version__}"
+                f"Current MLflow version: {mlflow.__version__}. "
+                f"Error: {e}"
```
Very hard for users registering to a remote server to debug unless this info is present

dbczumar committed:
Move mlflow module to import_namespace for conceptual clarity

Moves mlflow from local_namespace to import_namespace (globals) where it
conceptually belongs with other imports. This makes the code clearer since
import_namespace provides imports for name resolution while local_namespace
captures the output (created function). Functionally equivalent but more
maintainable.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/genai/scorers/scorer_utils.py:
```diff
@@ -82,59 +84,65 @@ def recreate_function(source: str, signature: str, func_name: str) -> Callable[.
+
+    # Import commonly used MLflow classes
     try:
-        # Parse the signature to build the function definition
```
Just remove try / catch block and unindenting so that the exception is made very clear / visible to users who register scorers to remove servers

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

harupy commented on mlflow/genai/scorers/scorer_utils.py:
```diff
@@ -82,59 +84,62 @@ def recreate_function(source: str, signature: str, func_name: str) -> Callable[.
+    func_def += indented_source
+
+    # Create a namespace with common MLflow imports that scorer functions might use
+    # Include mlflow module so type hints like "mlflow.entities.Trace" can be resolved
```
Not sure if this works but with `from __future__ import annotations`, we can skip runtime type hint evaluation.

dbczumar replied:
Added this

harupy reviewed (approved):
LGTM (https://github.com/mlflow/mlflow/pull/18043#discussion_r2394344275 might not work)

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Address PR feedback: add return type hint to test and use future annotations

- Add return type hint to test_function_with_mlflow_trace_type_hint signature
- Use 'from __future__ import annotations' in recreate_function to defer type hint evaluation
- This allows type hints with mlflow module references to work without runtime resolution

Signed-off-by: dbczumar <corey.zumar@databricks.com>
